### DISCUSSION TITLE
About headers error handling

### DISCUSSION THREAD

**Discussion by yiannis-kt** [fastapi/fastapi#7951] [Questions]
**Description**

When header is missing or does not pass some validation check (e.g. `min_length`) it gives "Internal Server Error" and no information about what is happening. In the logs it says something along the lines:

```
fastapi/routing.py", line 105, in app
    raise RequestValidationError(errors)
File "pydantic/error_wrappers.py", line 54, in pydantic.error_wrappers.ValidationError.__init__
TypeError: __init__() takes exactly 3 positional arguments (2 given)
```
What I would expect is to raise a BAD_REQUEST with information about what is going wrong.

**Example**

```
from fastapi import FastAPI, Header

app = FastAPI()


@app.get("/")
async def root(foo_bar: str = Header(..., min_length=10)):
    return {"message": foo_bar}
```

No header:
```
http GET http://localhost:8000/
```

Invalid header:
```
http GET http://localhost:8000/ foo-bar:foobar
```

**Additional context**

Python 3.7.4
fastapi 0.35.0

**Comment by euri10** [comment]
```
from fastapi import FastAPI, Header
from starlette.testclient import TestClient

app = FastAPI()


@app.get("/")
async def root(foobar: str = Header(..., min_length=10)):
    return {"message": foobar}


def test_header_lenght():
    with TestClient(app) as client:
        response = client.get("/", headers={"foobar": "0123456789"})
        assert response.status_code ==200

    with TestClient(app) as client:
        response = client.get("/", headers={"foobar": "012345678"})
        assert response.status_code ==422
```

**Comment by euri10** [comment]
I dont get why but it seems to me the underscore in your header is the root
of your concerns

On Tue, Oct 22, 2019 at 6:11 PM Yiannis Gatsoulis <notifications@github.com>
wrote:

> *Description*
>
> When header is missing or does not pass some min_length check it gives
> "Internal Server Error" and no information about what is happening. In the
> logs it says something along the lines:
>
>
> E.g.
>
>
> from fastapi import FastAPI, Header
>
> app = FastAPI()
>
> @app.get("/")
> async def root(foo_bar: str = Header(..., min_length=10)):
> return {"message": foo_bar}
>
>
>
> **Additional context**
> Add any other context or screenshots about the feature request here.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tiangolo/fastapi/issues/640?email_source=notifications&email_token=AAINSPX6VTABMLPSBFD2TXLQP4Q3JA5CNFSM4JDSBSAKYY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4HTRTFHQ>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINSPSO3CLBCPTLP4JKTI3QP4Q3JANCNFSM4JDSBSAA>
> .
>


-- 
benoit barthelet
http://pgp.mit.edu/pks/lookup?op=get&search=0xF150E01A72F6D2EE

**Comment by euri10** [comment]
ping @dmontagu ! :)

**Comment by euri10** [comment]
ok found it, you need to pass `convert_underscores=False` to your Headers should you want to use underscores in them, by default it translates underscores to dash

here's a passing test

```
import pytest
from fastapi import FastAPI, Header
from starlette.testclient import TestClient

app = FastAPI()


@app.get("/")
async def root(foo_bar: str = Header(..., min_length=10, convert_underscores=False)):
    return {"message": foo_bar}


def test_header_lenght():
    with TestClient(app) as client:
        response = client.get("/", headers={"foo_bar": "0123456789"})
        assert response.status_code ==200

    with TestClient(app) as client:
        response = client.get("/", headers={"foo_bar": "012345678"})
        assert response.status_code ==422
```

**Comment by yiannis-kt** [comment]
Many thanks @euri10. I actually need the opposite, i.e. `convert_underscores=True` as the header passed has a hyphen and this is resolved as such per description in the [docs](https://fastapi.tiangolo.com/tutorial/header-params/) (alternative is to use `alias` argument).

Problem was that I didn't realise I was using an older version of fastapi (0.35). Latest version (0.42) seems to working as expected.

**Comment by tiangolo** [comment]
Thanks for the help here @euri10 ! :clap: :bow: 

Thanks for reporting back and closing the issue @yiannis-kt  :+1:
